iceoryx_hoofs/include/iceoryx_hoofs/internal/posix_wrapper/mutex.hpp

```diff
@@ -17,81 +17,193 @@
-    pthread_mutex_t get_native_handle() const noexcept;
-
-    // NOLINTNEXTLINE(readability-identifier-naming) C++ STL code guidelines
-    pthread_mutex_t m_handle{};
+    Mutex(const Mutex&) = delete;
+    Mutex(Mutex&&) = delete;
+    Mutex& operator=(const Mutex&) = delete;
+    Mutex& operator=(Mutex&&) = delete;
+
+    /// @brief Locks the mutex.
+    /// @return When it fails it returns a MutexError describing the error.
+    ///         If the mutex is already locked it:
+    ///          * will be non blocking when the lock call comes from the same thread and the
+    ///              mutex type is MutexType::RECURSIVE
+    ///          * will be blocking when for all non MutexType::RECURSIVE
+    ///          * will be a MutexError::DEADLOCK_CONDITION when it is a
+    ///              MutexType::WITH_DEADLOCK_DETECTION
+
+    cxx::expected<MutexError> lock() noexcept;
+
+    /// @brief  Unlocks the mutex.
+    /// @return When it fails it returns a MutexError describing the error.
+    ///         If the mutex was locked by another thread and
+    ///         * MutexType::WITH_DEADLOCK_DETECTION it will return MutexError::NOT_OWNED_BY_THREAD
+    ///         * is not MutexType::WITH_DEADLOCK_DETECTION it may fail (depending on the platform)
+    ///         If the mutex was not locked it returns MutexType::NOT_OWNED_BY_THREAD
+    cxx::expected<MutexError> unlock() noexcept;
+
+    /// @brief Tries to lock the mutex.
+    /// @return If the lock was acquired MutexTryLock::LOCK_SUCCEEDED will be returned otherwise
+    ///         MutexTryLock::FAILED_TO_ACQUIRE_LOCK.
+    ///         If the lock is of MutexType::RECURSIVE the lock will also succeed.
+    ///         On failure it returns MutexError describing the failure.
+    cxx::expected<MutexTryLock, MutexError> try_lock() noexcept;
+
+  private:
+    Mutex() noexcept = default;
+
+  private:
+    friend class MutexBuilder;
+    friend class cxx::optional<Mutex>;
+
+    pthread_mutex_t m_handle;
+    bool m_isDescructable = true;
+};
+
+/// @todo iox-#1036 remove this, introduced to keep current API temporarily
+using mutex = Mutex;
+
+/// @brief Describes the type of mutex.
+enum class MutexType : int32_t
+{
+    /// @brief Behavior without error detection and multiple locks from within
+    ///        the same thread lead to deadlock
+    NORMAL = PTHREAD_MUTEX_NORMAL,
+
+    /// @brief Multiple locks from within the same thread do not lead to deadlock
+    ///        but one requires the same amount of unlocks to make the thread lockable
+    ///        from other threads
+    RECURSIVE = PTHREAD_MUTEX_RECURSIVE,
+
+    /// @brief Multiple locks from within the same thread will be detected and
+    ///        reported. It detects also when unlock is called from a different
+    ///        thread.
+    WITH_DEADLOCK_DETECTION = PTHREAD_MUTEX_ERRORCHECK,
+};
+
+/// @brief Describes the priority setting of the mutex.
+enum class MutexPriorityInheritance : int32_t
```

Comment:
I find the name a bit misleading. What about `MutexPrioritySetting`, `MutexPriorityAttribute`, `MutexProtocolAttribute`, `MutexProtocol`, ...?

Reply:
I would prefer the current name.